plutarch-extra/Plutarch/Extra/Ord.hs

```diff
@@ -0,0 +1,681 @@
+
+-- Removes all duplicates from a sorted list.
+--
+-- Only call this if you are _sure_ that the PLists are sorted according to the
+-- argument comparator!
+pnubUnsafe ::
+  forall (a :: PType) (s :: S).
+  Term s (PComparator a :--> PList a :--> PList a)
+pnubUnsafe = phoistAcyclic $
+  plam $ \cmp ->
+    precListLookahead (go cmp) (psingleton #) pnil
+  where
+    go ::
+      forall (s' :: S).
+      Term s' (PComparator a) ->
+      Term s' (a :--> PList a :--> PList a) ->
+      Term s' a ->
+      Term s' a ->
+      Term s' (PList a) ->
+      Term s' (PList a)
+    go cmp self h peek t =
+      pif
+        (pequateBy # cmp # h # peek)
+        (self # peek # t)
+        (pcons # h #$ self # peek # t)
+
+-- Breaks the argument into sorted chunks; currently, these range in size from 2
+-- to 4, with preference for larger chunks. To speed up the chunk sorting, we
+-- use sorting networks.
+pmergeStart_2_3_4 ::
+  forall (a :: PType) (ell :: PType -> PType) (s :: S).
+  (PElemConstraint ell a, PListLike ell) =>
+  Term s (PComparator a :--> ell a :--> PList (PList a))
+pmergeStart_2_3_4 = phoistAcyclic $
+  pfix #$ plam $ \self cmp ->
+    pmatchList pnil $
+      \_0 -> pmatchList (plist [psing _0]) $
+        \_1 -> pmatchList (plist [psort2 cmp _0 _1]) $
+          \_2 -> pmatchList (plist [psort3 cmp _0 _1 _2]) $
+            \_3 rest -> pcon $ PSCons (psort4 cmp _0 _1 _2 _3) (self # cmp # rest)
+  where
+    pmatchList ::
+      forall (r :: PType) (s' :: S).
+      Term s' r ->
+      (Term s' a -> Term s' (ell a) -> Term s' r) ->
+      Term s' (ell a) ->
+      Term s' r
+    pmatchList = flip pelimList
+
+-- Unhoisted Foldable lifter
+plist ::
+  forall (a :: PType) (f :: Type -> Type) (s :: S).
+  (Foldable f) =>
+  f (Term s a) ->
+  Term s (PList a)
+plist = foldr (\x -> pcon . PSCons x) (pcon PSNil)
+
+-- Unhoisted singleton maker
+psing ::
+  forall (a :: PType) (s :: S).
+  Term s a ->
+  Term s (PList a)
+psing x = plist [x]
+
+-- Two-item sorting network; basically a comparison.
+psort2 ::
+  forall (a :: PType) (s :: S).
+  Term s (PComparator a) ->
+  Term s a ->
+  Term s a ->
+  Term s (PList a)
+psort2 cmp _0 _1 =
+  pswap cmp _0 _1 $ \_0 _1 ->
+    plist [_0, _1]
+
+-- Three-item sorting network. The layers are:
+--
+
+-- * Arg 0 vs arg 2
+
+-- * Arg 0 vs arg 1
+
+-- * Arg 1 vs arg 2
+psort3 ::
+  forall (a :: PType) (s :: S).
+  Term s (PComparator a) ->
+  Term s a ->
+  Term s a ->
+  Term s a ->
+  Term s (PList a)
+psort3 cmp _0 _1 _2 =
+  pswap cmp _0 _2 $ \_0 _2 ->
+    pswap cmp _0 _1 $ \_0 _1 ->
+      pswap cmp _1 _2 $ \_1 _2 ->
+        plist [_0, _1, _2]
+
+-- Four-item sorting network. The layers are:
+--
+
+-- * Arg 0 vs arg 2 and arg 1 versus arg 3
+
+-- * Arg 0 vs arg 1 and arg 2 versus arg 3
+
+-- * Arg 1 versus arg 2
+
+--
+-- We perform the layer parallelism sequentially; it doesn't affect the
+-- semantics, it just makes us sad because it's slow.
+psort4 ::
```

Comment:
What's up with all this duplication? Just make a psortN

Reply:
The N would be a Haskell Integer

Reply:
It's heavily optimized by @cjay using some sorting network. You wouldn't get that performance without these.

Reply:
There is a huge amount of code duplication still. It seems like you can encode the sorting network as a table of numbers since the code is mostly the same. You can err for N more than 4.

Reply:
If you mean a table of numbers on the Haskell level, yeah it could probably be done while generating the same code on the script level, but that seems like a pointless exercise to me, unless you want to work with much larger sorting networks.
If you mean a table on the Plutarch-level, that seems useless if you want to sort arbitrary element types. If you can serialize into bytestrings then yeah, you could work with actual indices. But then you'd have to build the intermediate bytestrings. Maybe it would save significant script size on larger sorting networks, but probably with much higher execution cost.

Reply:
The former, it's much more succinct than adding a new function every time.

Reply:
Also, you can have a generic fallback for n > 4.